src/util/server_table.c

```diff
@@ -0,0 +1,177 @@
+      {
+        free_charbuf(&server_table.entries[i].server_id);
+      }
+      if (server_table.entries[i].cert.len != 0)
+      {
+        secure_free_charbuf(&server_table.entries[i].cert);
+      }
+    }
+  }
+  else
+  {
+    pelz_log(LOG_ERR, "Destroy Table Error");
+    return (1);
+  }
+
+  //Free the storage allocated for the hash table
+  free(server_table.entries);
+  pelz_log(LOG_DEBUG, "Server Table Destroy Function Complete");
+  return (0);
+}
+
+int server_table_delete(charbuf server_id)
+{
+  int index;
+
+  index = 0;
+  for (unsigned int i = 0; i < server_table.num_entries; i++)
+  {
+    if (cmp_charbuf(server_id, server_table.entries[i].server_id) == 0)
+    {
+      server_table.mem_size = server_table.mem_size -
+        ((server_table.entries[i].cert.len * sizeof(char)) + (server_table.entries[i].server_id.len * sizeof(char)) +
+        (2 * sizeof(size_t)));
+      free_charbuf(&server_table.entries[i].server_id);
+      secure_free_charbuf(&server_table.entries[i].cert);
+      index = i + 1;
+      break;
+    }
+  }
+  if (index == 0)
+  {
+    pelz_log(LOG_ERR, "Server ID not found.");
+    return (1);
+  }
+  else
+  {
+    for (unsigned int i = index; i < server_table.num_entries; i++)
+    {
+      server_table.entries[i - 1] = server_table.entries[i];
+    }
+    server_table.num_entries -= 1;
+    if ((server_table.entries =
+        (CertEntry *) realloc(server_table.entries, (server_table.num_entries) * sizeof(CertEntry))) == NULL)
+    {
+      pelz_log(LOG_ERR, "Server List Space Reallocation Error");
+      return (1);
+    }
+  }
+  return (0);
+}
+
+int server_table_add(charbuf server_id, uint64_t handle)
+{
+  CertEntry tmp_entry;
+  size_t max_mem_size;
+  charbuf tmpcert;
+  size_t data_size;
+  uint8_t *data;
+
+  max_mem_size = 1000000;
+
+  if (server_table.mem_size >= max_mem_size)
+  {
+    pelz_log(LOG_ERR, "Server Table memory allocation greater then specified limit.");
+    return (1);
+  }
+
+  tmp_entry.server_id = new_charbuf(server_id.len);
+  memcpy(tmp_entry.server_id.chars, server_id.chars, tmp_entry.server_id.len);
+  data_size = retrieve_from_unseal_table(handle, &data);
+  tmp_entry.cert = new_charbuf(data_size);
+  memcpy(tmp_entry.cert.chars, data, tmp_entry.cert.len);
+
+  if (!server_table_lookup(tmp_entry.server_id, &tmpcert))
+  {
+    if (cmp_charbuf(tmpcert, tmp_entry.cert) == 0)
+    {
+      pelz_log(LOG_DEBUG, "Cert already added.");
+      free_charbuf(&tmp_entry.server_id);
+      secure_free_charbuf(&tmp_entry.cert);
+      secure_free_charbuf(&tmpcert);
+      return (0);
+    }
+    else
+    {
+      pelz_log(LOG_ERR, "Cert entry and Server ID lookup do not match.");
+      free_charbuf(&tmp_entry.server_id);
+      secure_free_charbuf(&tmp_entry.cert);
+      secure_free_charbuf(&tmpcert);
+      return (1);
+    }
+  }
+
+  if ((server_table.entries =
+      (CertEntry *) realloc(server_table.entries, (server_table.num_entries + 1) * sizeof(CertEntry))) == NULL)
```

Comment:
This is probably not unique to here, but if ```realloc``` fails you've lost the pointer to the memory originally used for ```server_table_entries```. It's better to do something like this:
```
tmp_ptr = realloc(...)
if(tmp_ptr == NULL){
// error
}
else{
server_table.entries = tmp_ptr;
```